common/flatpak-transaction.c

```diff
@@ -4081,6 +4114,325 @@ flatpak_transaction_normalize_ops (FlatpakTransaction *self)
+          !priv->disable_related)
+        {
+          g_autoptr(GPtrArray) related = NULL;
+          if (!op_get_related (self, op, &related, error))
+            return FALSE;
+
+          for (i = 0; related && i < related->len; i++)
+            {
+              FlatpakRelated *rel = g_ptr_array_index (related, i);
+              if (rel->delete)
+                g_hash_table_add (newly_used_runtimes, g_strdup (rel->ref));
+            }
+        }
+
+      runtime_ref = op_get_runtime_ref (op);
+      if (runtime_ref != NULL)
+        {
+          g_autoptr(GPtrArray) runtime_related = NULL;
+          g_autofree char *runtime_remote = NULL;
+
+          if (op_type == FLATPAK_TRANSACTION_OPERATION_UNINSTALL)
+            {
+              runtime_related = find_related_from_deploy (self, runtime_ref, &runtime_remote);
+            }
+          else
+            {
+              FlatpakTransactionOperation *runtime_op = NULL;
+              runtime_op = flatpak_transaction_get_last_op_for_ref (self, runtime_ref);
+              if (runtime_op != NULL)
+                {
+                  runtime_remote = g_strdup (runtime_op->remote);
+                  if (!priv->disable_related && !op_get_related (self, runtime_op, &runtime_related, error))
+                    return FALSE;
+                }
+              else
+                runtime_related = find_related_from_deploy (self, runtime_ref, &runtime_remote);
+            }
+
+          if (op_type == FLATPAK_TRANSACTION_OPERATION_UNINSTALL)
+            {
+              g_hash_table_replace (maybe_unused_runtimes, g_strdup (runtime_ref), g_strdup (runtime_remote));
+              g_ptr_array_add (run_after_ops, op);
+            }
+          else
+            g_hash_table_add (newly_used_runtimes, g_strdup (runtime_ref));
+
+          if (!priv->disable_related)
+            {
+              for (i = 0; runtime_related && i < runtime_related->len; i++)
+                {
+                  FlatpakRelated *rel = g_ptr_array_index (runtime_related, i);
+
+                  if (op_type == FLATPAK_TRANSACTION_OPERATION_UNINSTALL)
+                    {
+                      if (rel->delete)
+                        g_hash_table_replace (maybe_unused_runtimes, g_strdup (rel->ref), g_strdup (runtime_remote));
+                    }
+                  else
+                    g_hash_table_add (newly_used_runtimes, g_strdup (rel->ref));
+                }
+            }
+        }
+
+      if (op_type == FLATPAK_TRANSACTION_OPERATION_UPDATE &&
+          dir_ref_is_installed (priv->dir, op->ref, NULL, &deploy_data))
+        {
+          g_autofree char *full_previous_runtime = NULL;
+          g_autofree char *prev_runtime_remote = NULL;
+          g_autoptr(GPtrArray) runtime_related = NULL;
+          g_autoptr(GPtrArray) ref_related = NULL;
+
+          g_ptr_array_add (run_after_ops, op);
+
+          /* The related refs of the op might now be unused */
```

Comment:
Couldn't the op->ref itself be unused now. For example, if a runtime which has nothing that depends on it is updated, and the new version is eol.